File: apps/mobile/lib/bjj.ts
import { apiRequest } from './apiRequest';
import { netFetch, SLOW_REQUEST_TIMEOUT_MS } from './authedFetch';
import type { TokenGetter } from './useAuthToken';

/**
 * BJJ rank, client half of `internal/modules/bjj`.
 *
 * `current` on {@link Standing} is DERIVED server-side from `promotions`, not
 * stored — see the backend's `StandingFrom`. This client never computes a
 * rank itself, for the same reason the server doesn't let a date decide it:
 * two independent derivations are two chances to disagree.
 */
export type Belt = 'white' | 'blue' | 'purple' | 'brown' | 'black';

/** Belts, in rank order — for a picker that doesn't hardcode the list again. */
export const BELTS: Belt[] = ['white', 'blue', 'purple', 'brown', 'black'];

export const MAX_STRIPES = 4;
export const MAX_DEGREE = 6;

export type Rank = {
  belt: Belt;
  stripes: number;
  /** Black-belt degrees. 0 on every other belt. */
  degree: number;
};

/** What the add/edit form sends — a Rank plus the promotion's own facts. */
export type PromotionInput = Rank & {
  /** "YYYY-MM-DD", or null when the athlete doesn't remember. */
  promoted_on: string | null;
  academy: string;
  instructor: string;
  note: string;
};

export type Promotion = PromotionInput & {
  id: string;
  /**
   * A presigned link that **expires** — see the backend handler. Never cache
   * it: a stored one is a broken image with extra steps. Absent when there is
   * no photo attached, or the environment has no object storage configured. A
   * promotion with no photo is exactly as valid a record as one with one; see
   * `promoted_on`'s own "undated promotion still establishes rank" reasoning.
   */
  photo_url?: string;
  created_at: string;
  updated_at: string;
};

export type Standing = {
  /** Null means no rank recorded — a real state, not a loading placeholder. */
  current: Rank | null;
  time_at_current_days: number | null;
  promotions: Promotion[];
};

/**
 * The promotion that awarded the rank the athlete currently holds.
 *
 * `Standing.current` is a Rank — belt, stripes, degree — and carries no
 * academy or date, because the server derives it as the *highest* recorded
 * rank rather than the latest row. The school and the date live on the
 * `Promotion` that granted it, so it has to be found by matching.
 *
 * Ties broken by the latest `promoted_on`, with undated promotions ranking
 * last: a rank entered twice (a correction, a re-entry) should show the date
 * the athlete most recently stated, and a dated record is better evidence than
 * an undated one whatever order they were typed in.
 *
 * Returns null when nothing matches — reachable, since the rank can be derived
 * from a promotion the athlete has since edited. The header then shows the
 * belt alone rather than inventing a school.
 */
export function awardingPromotion(standing: Standing): Promotion | null {
  const { current, promotions } = standing;
  if (!current) return null;
  const matches = promotions.filter(
    (p) => p.belt === current.belt && p.stripes === current.stripes && p.degree === current.degree,
  );
  if (matches.length === 0) return null;
  return matches.reduce((best, p) => {
    if (!p.promoted_on) return best;
    if (!best.promoted_on) return p;
    return p.promoted_on > best.promoted_on ? p : best;
  });
}

/**
 * "12 Mar 2024" — a promotion is a date you remember, not a timestamp.
 *
 * Short month deliberately: spelled out, "12 March 2024" overruns a third of a
 * phone's width and renders as "March 12, 20…", which loses the year — the one
 * part of a promotion date anybody actually quotes.
 */
export function formatAwardDate(day: string): string {
  // Parsed from parts rather than `new Date(day)`, which reads a bare
  // YYYY-MM-DD as UTC midnight and renders as the previous day west of
  // Greenwich — the same trap the plan module documents.
  const [y, m, d] = day.split('-').map(Number);
  if (!y || !m || !d) return day;
  return new Date(y, m - 1, d).toLocaleDateString(undefined, {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  });
}

export function getStanding(getToken: TokenGetter): Promise<Standing> {
  return apiRequest<Standing>(getToken, '/bjj/standing');
}

export function createPromotion(
  getToken: TokenGetter,
  input: PromotionInput,
): Promise<Promotion> {
  return apiRequest<Promotion>(getToken, '/bjj/promotions', {
    method: 'POST',
    body: JSON.stringify(input),
  });
}

export function updatePromotion(
  getToken: TokenGetter,
  id: string,
  input: PromotionInput,
): Promise<Promotion> {
  return apiRequest<Promotion>(getToken, `/bjj/promotions/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body: JSON.stringify(input),
  });
}

export function deletePromotion(getToken: TokenGetter, id: string): Promise<void> {
  return apiRequest<void>(getToken, `/bjj/promotions/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  });
}

type PromotionUploadTicket = {
  upload_url: string;
  content_type: string;
  max_bytes: number;
  expires_in: number;
  promotion: Promotion;
};

/**
 * Attach or replace a promotion's photo. Same two-step shape as
 * `body.ts`'s `uploadCheckinPhoto` and for the same reason: the bytes never
 * touch our API, only a short-lived signed URL does.
 *
 * **Unlike a check-in, the promotion must already exist** — there is no
 * date-keyed upsert to fall back on, so this can only be called with a real
 * id. A brand-new promotion is created first (`createPromotion`), and only
 * then can a photo be attached to the id it returns.
 *
 * The `promotion` on the returned ticket deliberately has an EMPTY
 * `photo_url` — see the backend handler's own comment: the key is
 * deterministic for this promotion, so presigning it here would resolve to
 * whatever was at that key before this upload finished. Callers that need to
 * display the fresh photo re-fetch (`getStanding`) after this resolves.
 */
export async function uploadPromotionPhoto(
  getToken: TokenGetter,
  id: string,
  localUri: string,
): Promise<Promotion> {
  const ticket = await apiRequest<PromotionUploadTicket>(
    getToken,
    `/bjj/promotions/${encodeURIComponent(id)}/photo`,
    { method: 'POST' },
  );

  const blob = await (await fetch(localUri)).blob();
  if (blob.size > ticket.max_bytes) {
    // The caller downscales before getting here; this is the backstop that
    // turns a silent storage rejection into a sentence.
    throw new Error(
      `That photo is ${Math.round(blob.size / 1024 / 1024)}MB — it needs to be under ${Math.round(
        ticket.max_bytes / 1024 / 1024,
      )}MB.`,
    );
  }

  const res = await netFetch(
    ticket.upload_url,
    {
      method: 'PUT',
      // Exactly the content type that was signed. Anything else is refused by
      // the signature, which is the point of signing it.
      headers: { 'Content-Type': ticket.content_type },
      body: blob,
    },
    // The slow budget, not the default: a multi-megabyte PUT to object
    // storage over whatever the gym's wifi is doing.
    { timeoutMs: SLOW_REQUEST_TIMEOUT_MS },
  );
  if (!res.ok) {
    throw new Error(`Couldn't upload that photo (${res.status}).`);
  }
  return ticket.promotion;
}

/**
 * The obvious next step from a current rank — one more stripe, or the next
 * belt once a stripe run is full.
 *
 * A suggestion for the add-promotion form to start from, not a value the
 * server ever sees or trusts: the athlete can change every field before
 * saving, and a jump (three stripes to a new belt, say) is exactly the case
 * this exists to make one tap instead of five.
 *
 * Black is degrees, not stripes — advancing past `MAX_DEGREE` has nowhere
 * further to go, so it holds rather than wrapping to white.
 */
export function nextRank(current: Rank): Rank {
  if (current.belt === 'black') {
    return { belt: 'black', stripes: 0, degree: Math.min(current.degree + 1, MAX_DEGREE) };
  }
  if (current.stripes < MAX_STRIPES) {
    return { belt: current.belt, stripes: current.stripes + 1, degree: 0 };
  }
  const next = BELTS[Math.min(BELTS.indexOf(current.belt) + 1, BELTS.length - 1)];
  return { belt: next, stripes: 0, degree: 0 };
}

/**
 * "3 years", "6 months", "12 days" — the coarsest unit that doesn't round to
 * zero. Matches how a grappler actually states time at a belt; nobody says
 * "1,097 days".
 */
export function describeTimeAtBelt(days: number): string {
  const years = Math.floor(days / 365);
  if (years >= 1) return `${years} ${years === 1 ? 'year' : 'years'}`;
  const months = Math.floor(days / 30);
  if (months >= 1) return `${months} ${months === 1 ? 'month' : 'months'}`;
  return `${days} ${days === 1 ? 'day' : 'days'}`;
}

File: backend/internal/modules/bjj/bjj.go
// Package bjj is the jiu-jitsu-specific athlete record: rank, and the
// promotion history that rank is derived from.
//
// It lives here rather than on `profile` because profile is the account-level
// record four disciplines share, and a belt is meaningless to three of them —
// see the note at the top of profile.go, which called this split before there
// was anything to put in it.
package bjj

import (
	"context"
	"errors"
	"time"
)

var (
	ErrNotFound     = errors.New("bjj: not found")
	ErrInvalidInput = errors.New("bjj: invalid input")
)

// Belt is an adult IBJJF belt.
//
// Kids belts (grey/yellow/orange/green, each in three variants) and the
// coral/red belts are deliberately absent. Adding either is an edit to this
// slice plus a rendering branch in the clients — there is no CHECK constraint
// in the database precisely so neither needs a migration.
type Belt string

const (
	White  Belt = "white"
	Blue   Belt = "blue"
	Purple Belt = "purple"
	Brown  Belt = "brown"
	Black  Belt = "black"
)

// belts, in order. The ORDER is the load-bearing part: it decides which
// promotion is current, so the slice is the single source for both validity
// and comparison.
var belts = []Belt{White, Blue, Purple, Brown, Black}

// MaxStripes is 4 on every belt. A fifth stripe is what promotion to the next
// belt means, so 5 is not a rank this can represent.
const MaxStripes = 4

// MaxDegree is 6. Degrees above that (7th–9th are the coral and red belts)
// are a different belt colour, not a further degree of black.
const MaxDegree = 6

// Rank is a point in the belt system.
type Rank struct {
	Belt    Belt `json:"belt"`
	Stripes int  `json:"stripes"`
	// Degree is black-belt degrees. 0 on every other belt, and 0 on a black
	// belt that has none — both render the same, so there is nothing to
	// distinguish and no reason for a pointer.
	Degree int `json:"degree"`
}

// Order returns a value that sorts ranks correctly, or false for an unknown
// belt.
//
// **This is what "current rank" means, and it is not the most recent date.**
// Dates are optional (plenty of people do not remember when they got their
// blue belt) and are entered by hand, so ordering by them makes the current
// belt a function of data-entry care. Rank in BJJ is monotonic — nobody is
// demoted — so the highest rank recorded IS the current one, and that holds
// whether the promotions were entered in order, backwards, or with no dates
// at all.
func (r Rank) Order() (int, bool) {
	for i, b := range belts {
		if b == r.Belt {
			// Stripes and degrees are within-belt, so they cannot outrank a
			// belt change: white/4 must still sort below blue/0. Multiplying
			// by the maxima+1 keeps each level strictly dominant over the one
			// below it.
			return i*(MaxStripes+1)*(MaxDegree+1) +
				r.Stripes*(MaxDegree+1) +
				r.Degree, true
		}
	}
	return 0, false
}

// Validate reports whether this is a rank the system can represent.
func (r Rank) Validate() error {
	if _, ok := r.Order(); !ok {
		return ErrInvalidInput
	}
	if r.Stripes < 0 || r.Stripes > MaxStripes {
		return ErrInvalidInput
	}
	if r.Degree < 0 || r.Degree > MaxDegree {
		return ErrInvalidInput
	}
	// A degree on a coloured belt is not a thing. Rejecting it rather than
	// silently zeroing it: the client sent something it believed, and quietly
	// dropping a field is how a UI ends up displaying a value the server
	// never stored.
	if r.Degree > 0 && r.Belt != Black {
		return ErrInvalidInput
	}
	return nil
}

// Belts lists the belts in order, for a client that wants to render a picker
// without hardcoding the list a fourth time.
func Belts() []Belt {
	out := make([]Belt, len(belts))
	copy(out, belts)
	return out
}

// Promotion is one rank event: what was awarded, when, where and by whom.
type Promotion struct {
	ID     string `json:"id"`
	UserID string `json:"-"`
	Rank   `json:""`
	// PromotedOn is "YYYY-MM-DD", or nil when the athlete does not remember.
	// An undated promotion still establishes rank — refusing it to protect
	// the metadata would lose the fact itself.
	PromotedOn *string   `json:"promoted_on"`
	Academy    string    `json:"academy"`
	Instructor string    `json:"instructor"`
	Note       string    `json:"note"`
	CreatedAt  time.Time `json:"created_at"`
	UpdatedAt  time.Time `json:"updated_at"`
}

// Standing is the whole BJJ rank picture for one athlete.
type Standing struct {
	// Current is the highest recorded rank, or nil if nothing is recorded.
	// Nil is a real state and clients must render it — a new account has no
	// belt, and guessing white for them would put a rank on someone who has
	// never trained.
	Current *Rank `json:"current"`
	// TimeAtCurrentDays is days since the promotion that set the current
	// rank, or nil when that promotion has no date. Nil rather than 0: "I
	// don't know" and "today" are different answers.
	TimeAtCurrentDays *int        `json:"time_at_current_days"`
	Promotions        []Promotion `json:"promotions"`
}

// Repository is the persistence port.
type Repository interface {
	ListPromotions(ctx context.Context, userID string) ([]Promotion, error)
	CreatePromotion(ctx context.Context, p Promotion) (Promotion, error)
	UpdatePromotion(ctx context.Context, p Promotion) (Promotion, error)
	DeletePromotion(ctx context.Context, userID, id string) error
}

// StandingFrom derives the whole picture from a promotion list.
//
// A pure function over the rows, so the derivation is testable without a
// database and cannot differ between the read path and any future writer.
func StandingFrom(promotions []Promotion, now time.Time) Standing {
	// Normalised here, not left to the caller. The repository already returns
	// a non-nil empty slice, but this function is what builds the response —
	// so a nil from anywhere else would marshal as `null` and force a null
	// check on every client that iterates it.
	if promotions == nil {
		promotions = []Promotion{}
	}
	s := Standing{Promotions: promotions}
	if len(promotions) == 0 {
		return s
	}

	best := -1
	var current *Promotion
	for i := range promotions {
		o, ok := promotions[i].Order()
		// An unknown belt is skipped rather than sorted as zero. A row
		// written by a newer build (coral, say) must not be read as
		// something below white and silently ignored in favour of a real
		// white belt — better to act as though the row is not there.
		if !ok {
			continue
		}
		if o > best {
			best, current = o, &promotions[i]
		}
	}
	if current == nil {
		return s
	}

	rank := current.Rank
	s.Current = &rank
	if current.PromotedOn != nil {
		if d, err := time.Parse("2006-01-02", *current.PromotedOn); err == nil {
			days := int(now.Sub(d).Hours() / 24)
			// A promotion dated in the future is someone's typo. Reporting a
			// negative time-at-belt would be worse than reporting none.
			if days >= 0 {
				s.TimeAtCurrentDays = &days
			}
		}
	}
	return s
}
